traits/trait_types.py

```diff
@@ -2774,6 +2774,178 @@ def items_event(self):
 AdaptMap = {"no": 0, "yes": 1, "default": 2}
 
 
+class Map(TraitType):
+    """ Checks that the value assigned to a trait attribute is a key of a
+        specified dictionary, and also assigns the dictionary value
+        corresponding to that key to a *shadow* attribute.
+
+        A trait attribute of type Map is called a *mapped* trait
+        attribute. In practice, this means that the resulting object actually
+        contains two attributes: one whose value is a key of the Map
+        dictionary, and the other whose value is the corresponding value of the
+        Map dictionary. The name of the shadow attribute is simply the base
+        attribute name with an underscore ('_') appended. Mapped trait
+        attributes can be used to allow a variety of user-friendly input values
+        to be mapped to a set of internal, program-friendly values.
+
+        Example
+        -------
+
+        The following example defines a ``Person`` class::
+
+            >>>class Person(HasTraits):
+            ...    married = Map({'yes': 1, 'no': 0 })
+            >>>
+            >>>bob = Person()
+            >>>print bob.married
+            yes
+            >>>print bob.married_
+            1
+
+        In this example, the default value of the ``married`` attribute of the
+        Person class is 'yes'. Because this attribute is defined using
+        PrefixList, instances of Person have another attribute,
+        ``married_``, whose default value is 1, the dictionary value
+        corresponding to the key 'yes'.
+
+        Parameters
+        ----------
+        map : dict
+            A dictionary whose keys are valid values for the trait attribute,
+            and whose corresponding values are the values for the shadow
+            trait attribute.
+
+        Attributes
+        ----------
+        map : dict
+            A dictionary whose keys are valid values for the trait attribute,
+            and whose corresponding values are the values for the shadow
+            trait attribute.
+    """
+
+    is_mapped = True
+
+    def __init__(self, map, **metadata):
+
+        self.map = map
+        self.fast_validate = (ValidateTrait.map, map)
+
+        default_value = metadata.pop("default_value", None)
+
+        if default_value is not None:
+            try:
+                default_value = self.validate(None, None, default_value)
+            except TraitError:
+                raise ValueError("The default value for Map trait must be key"
+                                 " in the map.")
+
+        super().__init__(default_value, **metadata)
+
+    def validate(self, object, name, value):
+        try:
+            if value in self.map:
+                return value
+        except:
```

Comment:
Changed to `except Exception`

Reply:
Since this is new code and we don't have backwards compatibility constraints, let's use something more targeted that `except Exception`.

Reply:
General rule: `except Exception`, `except BaseException` and bare `except` are things you want to use rarely: the main examples are (i) places where you plan to re-raise the exception (possibly after logging), or (ii) across major application boundaries (e.g., between model/algorithm and UI, where you want to prevent an exception in the model from bringing down the UI, or when running code in a background thread or process). Outside of these cases, the big issue with these statements is that they risk hiding / silencing other coding errors, and make future debugging more difficult as a result.